src/tools/clippy/clippy_lints/src/significant_drop_in_scrutinee.rs

```diff
@@ -0,0 +1,206 @@
+    ///
+    ///     let is_foo = mutex.lock().unwrap().foo();
+    ///     match is_foo {
+    ///         true => {
+    ///             mutex.lock().unwrap().bar();
+    ///         }
+    ///         false => {}
+    ///     };
+    ///
+    ///     println!("All done!");
+    /// }
+    /// ```
+    #[clippy::version = "1.60.0"]
+    pub SIGNIFICANT_DROP_IN_SCRUTINEE,
+    nursery,
+    "warns when a temporary of a type with a drop with a significant side-effect might have a surprising lifetime"
+}
+
+declare_lint_pass!(SignificantDropInScrutinee => [SIGNIFICANT_DROP_IN_SCRUTINEE]);
+
+impl<'tcx> LateLintPass<'tcx> for SignificantDropInScrutinee {
+    fn check_expr(&mut self, context: &LateContext<'tcx>, expression: &'tcx Expr<'tcx>) {
+        if has_significant_drop_in_scrutinee(expression, context) {
+            span_lint_and_help(
+                context,
+                SIGNIFICANT_DROP_IN_SCRUTINEE,
+                expression.span,
+                "The lifetimes of temporaries in the match scrutinee will last until the end of \
+                 the match expression, but one or more of them has a drop with a significant side-effect.",
+                None,
+                "The lifetimes of temporaries in a match scrutinee will last until the end of \
+                the match expression, but one of the temporaries in this match scrutinee has a drop \
+                with a significant side-effect (such as unlocking a mutex). Consider calling the \
+                function(s) outside of the match expression and using the result in the scrutinee.",
+            );
+        }
+    }
+}
+
+/// If the expression is an ExprKind::Match, check if the scrutinee has a significant drop that may
+/// have a surprising lifetime.
+fn has_significant_drop_in_scrutinee<'tcx>(expression: &'tcx Expr<'tcx>, context: &LateContext<'tcx>) -> bool {
+    match expression.kind {
+        ExprKind::Match(match_expr, _, _source) => is_expr_with_significant_drop(&match_expr, context, true),
+        _ => false,
+    }
+}
+
+/// Recursively check the expressions in the scrutinee to see if any of them have a return type that
+/// has `has_significant_drop` attribute. If one is found and it *is* the last method call, further
+/// check if there are previous method calls to determine if an arm is borrowing or taking ownership
+/// of the returned type. If there are subsequent method calls, we can infer that the lifetime of
+/// the returned type is *potentially* surprising, so return true. In all other cases, return false.
+fn is_expr_with_significant_drop<'a, 'tcx>(
+    expression: &Expr<'a>,
+    context: &LateContext<'tcx>,
+    is_last_method_call: bool,
+) -> bool {
+    if is_last_method_call || !is_expr_with_significant_drop_attribute(expression, context) {
+        match expression.peel_drop_temps().kind {
+            ExprKind::MethodCall(_path, exprs, _source) => is_expr_with_significant_drop(&exprs[0], context, false),
```

Comment:
I'm not sure if I'm doing something wrong, but I couldn't get this to work. It was trying to move `expr` (and then produce a compile error because it was moving from a reference). Taking a ref to `expression.peel_drop_temps().kind` had the same compile error. Is there another approach I should be taking?